### DISCUSSION TITLE
How can I match ModelProducer series with LineProvider series

### DISCUSSION THREAD

**Discussion by dmytroreutov** [patrykandpatrick/vico#907] [Questions]
### Question

Hello. According logic in my app, chart should be able to draw 3 styles of lines.  
1) blue for download test dataset
2) green for upload test dataset
3) red for a state of failure one or both tests

I've declared these styles in styles.xml like:

```
    <style name="ChartLineStyleDownload">
        <item name="color">@color/speedtest_chart_line_download_start</item>
        <item name="thickness">3dp</item>
        <item name="curvature">100%</item>
    </style>

    <style name="ChartLineStyleUpload">
        <item name="color">@color/speedtest_chart_line_upload_start</item>
        <item name="thickness">3dp</item>
        <item name="curvature">100%</item>
    </style>

    <style name="ChartLineStyleFailure">
        <item name="color">@color/speedtest_chart_line_failure</item>
        <item name="thickness">3dp</item>
        <item name="curvature">100%</item>
    </style>

    <style name="ChartLineLayerStyle">
        <item name="line1Style">@style/ChartLineStyleDownload</item>
        <item name="line2Style">@style/ChartLineStyleUpload</item>
        <item name="line3Style">@style/ChartLineStyleFailure</item>
    </style>
```

assigned this style to chart in my layout like:

`app:lineLayerStyle="@style/ChartLineLayerStyle"`

then in code i'm handling conditions like:

```
        lifecycleScope.launch {
            modelProducer.runTransaction {
                if (isIdle) return@runTransaction

                lineSeries {
                    if (isProgress || isDone) {
                        series(downloadSet)
                        series(uploadSet)
                    }
                    if (isFailure || status == NetworkStatus.Offline) {
                        series(viewModel.errorSet)
                    }
                }
            }
        }
```
`isIdle` - graph should be cleared of lines, thats why i do `return`
`isProgress || isDone` -  one or both lines `style1`/`style2` representing test should be displayed
`isFailure || status == NetworkStatus.Offline` - stub line should be drawn in `style3`

Unfortunately, `errorSet` is drawn in `style1` instead of `style3`. I've seen recommendations to use `LineProvider` to dynamically modify each line. But I don't see any examples how can I match `LineProvider`  series to `ModelProducer` series, to pick defined style for condition.

### Vico version(s)

2.0.0-alpha.28

### UI framework(s)

Views

**Comment by Gowsky** [comment]
Hello, @dmytroreutov. Please see [here](https://patrykandpatrick.com/vico/wiki/cartesian-charts/data/#asynchrony-and-extras) for information on how to use `ExtraStore.Key`s. You can use them to store additional information helping you identify the dataset type when running a `Transaction`. `LineProvider` interface allows you to retrieve the information with `extraStore` parameter, so you can return the desired style depending on the dataset type.

**Reply by dmytroreutov** [reply]
Hi! Thank you for your response.
Unfortunately, I didn't get it.

When should I call `extras {}` ?
1) after entire` lineSeries {}` block
2) after each `series()`
3) before any of that?

Also, I've noticed `LineCartesianLayer.LineProvider#getLine` has `seriesIndex: Int` as parameter. 
Is it possible somehow add series like `series(seriesIndex, data)`?

`ExtraStore` approach is kind of complicated and I cannot understand how to use it properly.
Should I predefine 3 keys (for each line) and assign some value for them in `runTransaction {}` block? 
If that, what exactly is better to store there? Entire datasets? And how I can properly access it from `LineProvider`?

**Reply by dmytroreutov** [reply]
Or should I store each line index in `ExtraStore` and then trough `LineCartesianLayer.LineProvider#getLine` obtain existing indices and create line based on each index?

**Reply by dmytroreutov** [reply]
It would be perfect to use only 

```
<style name="ChartLineLayerStyle">
        <item name="line1Style">@style/ChartLineStyleDownload</item>
        <item name="line2Style">@style/ChartLineStyleUpload</item>
        <item name="line3Style">@style/ChartLineStyleFailure</item>
</style>
```

and add series with index 0,1,2 (or 1,2,3), to avoid customisation via `ExtraStore`

**Reply by Gowsky** [reply]
@dmytroreutov, I'm reposting this comment with edits to better clarify a few things.

>When should I call extras {} ?
>1. after entire lineSeries {} block
>2. after each series()
>3. before any of that?

It doesn't matter when you call it.

>Also, I've noticed LineCartesianLayer.LineProvider#getLine has seriesIndex: Int as parameter.
Is it possible somehow add series like series(seriesIndex, data)?

The index of a series is determined automatically based on the order of the `series` calls. The ability to specify any index whatsoever would be in contradiction with the definition of an index.

>ExtraStore approach is kind of complicated and I cannot understand how to use it properly.
Should I predefine 3 keys (for each line) and assign some value for them in runTransaction {} block?
If that, what exactly is better to store there? Entire datasets?

`ExtraStore` is a typed `Map` — it can store anything. You should save whatever enables you to determine the desired `Line` based on the series index. You don't need the entire dataset to do so. Based on the information provided, there are two options:

- The series at index 0 corresponds to download (`style1`), and the series at index 1 corresponds to upload (`style2`).
- The series at index 0 corresponds to error (`style3`).

Therefore, even a boolean would suffice, though it wouldn't be the most scalable solution. A better approach may be to create an enum with `Download`, `Upload`, and `Error` members and save a list of members of this enum as an extra (either `[Download, Upload]` or `[Error]`). Then, in `LineProvider.getLine`, you can use `seriesIndex` to read from this list. That will tell you what the series in question represents, enabling you to return the right `Line`. The choice of solution is yours.

>And how I can properly access it from LineProvider?

`LineProvider.getLine` has an `ExtraStore` parameter.

>Or should I store each line index in ExtraStore and then trough LineCartesianLayer.LineProvider#getLine obtain existing indices and create line based on each index?

You should have the three `Line`s predefined and choose the right one from `LineProvider.getLine` based on `seriesIndex` and the extra. More details above.

>and add series with index 0,1,2 (or 1,2,3), to avoid customisation via ExtraStore

This is once again in contradiction with the definition of an index. I understand that the broader point is the ability to add keys to series, but that would be only slightly simpler than the extra-based solution proposed above, and it would be less flexible.
